design-proposals/klip-38-variable-substitution.md

```diff
@@ -0,0 +1,302 @@
+```
+Valid variables names start with a letter or underscore (\_) followed by zero or more alphanumeric characters or underscores.
+
+All variable values need to be wrapped using single quotes. All variables are stored as string values in memory, and when
+replaced, the value without the quotes will be used. ksqlDB parsing will be in charge of determining the format at that
+time.
+
+The reason not to remove single-quotes from values is to allow spaces, and to leave the syntax open for using data types in the future if required.  
+Also, having decide the format at run-time (i.e. `define n = 3.34238`) is tricky as we could lose precision in case of decimals. Better stay with Strings for now.
+
+Define a new variable:
+```
+ksql> DEFINE replicas = '3';
+ksql> DEFINE partitions = '5';
+```
+
+Print a defined variable value:
+```
+ksql> DEFINE replicas;
+DEFINE replicas   = '3'
+```
+
+List all defined variables:
+```
+ksql> DEFINE;
+DEFINE replicas   = '3'
+DEFINE partitions = '5'
+```
+
+#### Deleting substitution variables
+
+The `UNDEFINE` command deletes the variable.
+
+Syntax:
+```
+UNDEFINE name;
+```
+
+### 2. Scope
+
+This variable has a session scope.
+
+Variables will live during the session of a ksqlDB connection, then destroyed when the connection is closed. Only the user
+in that session can interact with the defined substitution variables. No other users can interact with other users' variables.
+
+In CLI, the session is opened when starting the CLI with the `ksql` command, and destroyed when typing `exit`.
+
+In Java API client, the session is opened when `Client.create()` is called, and closed when `Client.close()` is called. Variables
+may be defined using a client method instead of a new syntax. This is to ensure users know the scope starts after the `create()` call.
+
+TBD: Define a Java API method for variable substitution.
+
+In headless, the session is opened when starting the ksqlDB server with a SQL script. The variables stay alive until the server is
+killed.
+
+Variables are not persisted anywhere. They are living in memory as long as the session is not terminated.
+
+#### RUN SCRIPT command
+
+If a `RUN SCRIPT` command is executed in a CLI session, then any variable defined there will be local to the execution context of the
+script. The variables will override any variable previously defined in the session. When the script is completed, all local variables
+defined will be destroyed. This ensures users do not override their session variables by accident when running `RUN SCRIPT` commands.
+
+#### The SOURCE command
+
+To allow users load variables and keep their scope in the session, then the new `SOURCE` command will be added. This will override
+any previously defined variables, and will be destroyed until the user leaves the session.
+
+Syntax:
+```
+SOURCE '<script_file>';
+```
+
+The `SOURCE` command will work like a `RUN SCRIPT`. If other SQL statements are in the file, then those statements will also be
```

Comment:
What's the benefit in having `SOURCE` allow users to execute statements similar to `RUN SCRIPT`? It seems confusing to have two ways to do this with only a very subtle distinction (variables set in `SOURCE` persist into the sessions outside the scope of the command while those for `RUN SCRIPT` don't). Can we have `SOURCE` _only_ load variables from a file, and then ask users to use `RUN SCRIPT` afterwards if they want to also load statements?

Reply:
We could that too. I'm not sure how users will use their scripts, though. Are they willing to split their scripts between vars and commands?

I just though of using the bash behavior:
```
$ cat /tmp/s1
export ID=1
echo id=$ID

$ cat /tmp/s2
export ID=2
echo id=$ID

$ source /tmp/s1
id=1
$ bash /tmp/s2
id=2

// Uses the ID set in by the source command
$ echo $ID
1
```

Mysql uses SOURCE too, and it allows variables and commands.
https://dev.mysql.com/doc/refman/5.7/en/mysql-batch-commands.html

Reply:
Does `MySQL` also have a `RUN SCRIPT`? I tend to agree with Victoria that this seems a little confusing.